### PR TITLE
support alfred 4 and 3

### PR DESCRIPTION
**Pull request by karlsander**
This PR adds support for Alfred 4, which completely drops the version postfix while keeping support for v3.

It's not exactly elegant, but supporting 2 different paths just isn't I guess. Here's an alternative, [simple PR](https://github.com/SamVerschueren/resolve-alfred-prefs/pull/6) that just drops v3 support.

### CODE REVIEW

**Comment by LitoMore** [comment]
I think we could add a parameter to the method like `resolveAlfredPrefs(options)`.

Then we can use this module in [alfy](https://github.com/sindresorhus/alfy) twice for checking both Alfred 3 and Alfred 4.

**Comment by LitoMore** [comment]
@SamVerschueren @sindresorhus Any comments with this discussion?

**Comment by LitoMore** [comment]
More details for my idea:

#### `resolveAlfredPrefs(options)`

##### `options.settingsPath`

Default to `/Library/Preferences/com.runningwithcrayons.Alfred-Preferences.plist`, the latest version of Alfred.

---

When we using this module in `alfy`, we should check both Alfred 3 and Alfred 4.

```javascript
const prefsLatest = await resolveAlfredPrefs();
const prefs3 = await resolveAlfredPrefs({
  settingsPath: '/Library/Preferences/com.runningwithcrayons.Alfred-Preferences.plist'
})

// Doing symlink for Alfred 3 and Alfred 4
// ...
```

**Comment by deanishe** [comment]
For Alfred 4+, the correct way to get the path to Alfred's preferences bundle is to read the `current` value from `~/Library/Application Support/Alfred/prefs.json`.

The file was added specifically for this purpose.

**Comment by LitoMore** [comment]
I will make a new pull request to add support for both 3 and 4. Then I will ask Sindre to help me review my pull request.

**Comment by sindresorhus** [comment]
Alright. I'm ok with supporting Alfred 3 and 4 using the method @deanishe has suggested.

**Comment by LitoMore** [comment]
@sindresorhus Here are my ideas about the API design, could you give me some advice?

### Idea 1

```javascript
const resolveAlfredRefs = require('resolve-alfred-refs');

(async () => {
  console.log(await resolveAlfredRefs());
  //=> {alfred3: '/path/to/alfred-3-prefs', alfred4: undefined}
})();
```

If Alfred 3/4 does not exist, the value of `alfred4` will be `undefined`. This function will never throw any error.

### Idea 2

```javascript
const {resolveAlfred3Prefs, resolveAlfred4Prefs} = require('resolve-alfred-refs');

(async () => {
  const alfred3 = await resolveAlfred3Prefs();
  console.log(alfred3);
  //=> '/path/to/alfred-3-prefs'

  const alfred4 = await resolveAlfred4Prefs();
  console.log(alfred4);
  //=> '/path/to/alfred-4-prefs'
})();
```

If Alfred 3/4 does not exist or the user doesn't have permission to access, the function will throw an error.

### Idea 3

Make a new module for Alfred 4:

I created a new package [resolve-alfred-4-refs](https://github.com/LitoMore/resolve-alfred-4-prefs) (not published to npm yet) to resolve the preferences file for Alfred 4.

We can use this package in [alfred-link](https://github.com/SamVerschueren/alfred-link) directly.

**Comment by sindresorhus** [comment]
I would prefer:

```js
const resolveAlfredRefs = require('resolve-alfred-refs');

(async () => {
  console.log(await resolveAlfredRefs());
  //=> {isAlfred4: true, path: '/path/to/alfred-4-prefs'}
})();
```

**Comment by deanishe** [comment]
> Make a new module for Alfred 4

I don't think this is a good idea, at least not an Alfred 4-specific module. Alfred 4+ will all use the same directories (i.e. no more "Alfred X" and no more workflows breaking with every major update). It should mostly be a case of one (legacy) method for Alfred 3 and one new one for every later version.

You should use `prefs.json` first (which should work with Alfred 4, Alfred 5, etc.) and only fall back to the old method for Alfred 3 if `prefs.json` doesn't exist.

To the same ends, any JXA calls should use `Application('com.runningwithcrayons.Alfred')`, not `Application('Alfred 4')`.

**Comment by SamVerschueren** [comment]
I agree with what @sindresorhus commented. Something in line with

```js
const resolveAlfredRefs = require('resolve-alfred-refs');

(async () => {
  console.log(await resolveAlfredRefs());
  //=> {isAlfred4: true, path: '/path/to/alfred-4-prefs'}
})();
```

Or `{version: '4', path: '/path/to/alfred-4-prefs'}`. This way, we have an upgrade path if v5 etc. comes out.

**Comment by sindresorhus** [comment]
@SamVerschueren Sounds good. I don't think we even need the version number, I guess we can have it just in case.

**Comment by deanishe** [comment]
> I don't think we even need the version number, I guess we can have it just in case.

You need the version number to abort an update to an incompatible version, don't you? Or is it already too late to do that when this module is called?

**Comment by SamVerschueren** [comment]
That could definitely be added to [alfred-link](https://github.com/SamVerschueren/alfred-link). But it also means that every workflow should define the compatible Alfred version.

**Comment by deanishe** [comment]
> But it also means that every workflow should define the compatible Alfred version.

I'm not familiar with the way you have this whole workflows-via-npm business working. The main issue to avoid, imo, is updating workflows to newer, incompatible versions.

If updating a workflow requires users to run `npm` themselves, I don't think it's a problem. (Well, it's *their* problem.)

You just want to avoid workflows overwriting themselves with incompatible versions.

**Comment by LitoMore** [comment]
> You should use prefs.json first (which should work with Alfred 4, Alfred 5, etc.) and only fall back to the old method for Alfred 3 if prefs.json doesn't exist. (@deanishe)

Good point.

So I think the API returns should be:

```javascript
// If Alfred 4 or later
{path: '/path/to/prefs'}

// If Alfred 3
{version: '3', path: '/path/to/prefs'}
```

File: index.js [hunk 1]
```diff
@@ -7,17 +7,22 @@ const bplistParser = require('bplist-parser');
...
+const settingsLatest = path.join(userHome, '/Library/Preferences/com.runningwithcrayons.Alfred-Preferences.plist');
 
 module.exports = async () => {
-	const exists = await pathExists(settings);
+	const latestExists = await pathExists(settingsLatest);
+	const v3Exists = await pathExists(settingsV3);
 
-	if (!exists) {
-		throw new Error(`Alfred preferences not found at location ${settings}`);
+	if (!(latestExists || v3Exists)) {
```
Comment (SamVerschueren) [inline index.js:17]: IMO it's easier to read if you did it like this

```suggestion
	if (!latestExists && !v3Exists) {
```

File: test/test.js [hunk 2]
```diff
@@ -6,13 +6,13 @@ import userHome from 'user-home';
...
 
-const settings = path.join(userHome, '/Library/Preferences/com.runningwithcrayons.Alfred-Preferences-3.plist');
+const settings = path.join(userHome, '/Library/Preferences/com.runningwithcrayons.Alfred-Preferences.plist');
 
 const mv = (src, dest) => pify(fs.rename)(src, dest).catch(() => { });
 
 test.before(async () => {
 	await mv(settings, `${settings}.back`);
-	await cp(path.join(__dirname, 'fixtures/com.runningwithcrayons.Alfred-Preferences-3.plist'), settings);
+	await cp(path.join(__dirname, 'fixtures/com.runningwithcrayons.Alfred-Preferences.plist'), settings);
```
Comment (SamVerschueren) [inline test/test.js:15]: We should test both Alfred 3 and latest.

So maybe we can make it a test [macro](https://github.com/avajs/ava/blob/master/docs/01-writing-tests.md#reusing-test-logic-through-macros)?

```js
function macro(t, filename, expected) {
	t.is(eval(input), expected);
}

macro.title = (providedTitle, filename) => `resolves \`${filename}\``;

test(macro, 'Alfred-Preferences.plist', path.join(userHome, 'Documents/alfred/Alfred.alfredpreferences'));
test(macro, 'Alfred-Preferences-3.plist', path.join(userHome, 'Documents/alfred/Alfred.alfredpreferences'));
```

**Review by SamVerschueren — CHANGES_REQUESTED** [review]

**Comment by SamVerschueren** [comment]
> So I think the API returns should be:
> 
> ```js
> // If Alfred 4 or later
> {path: '/path/to/prefs'}
> 
> // If Alfred 3
> {version: '3', path: '/path/to/prefs'}
> ```

It would be nice if we could retrieve the version from the `plist` file to always provide it. That way, we could enforce a `engines.alfred` setting in `package.json` so it fails to install if you are installing a Alfred 4 workflow for Alfred 3.

So something like this would be better

```js
// If Alfred 3
{version: 3, path: '/path/to/prefs'}

// If Alfred 4
{version: 4, path: '/path/to/prefs'}

// If Alfred 5
{version: 5, path: '/path/to/prefs'}
```

(Not sure if versions are numeric though).

If that's not possible, the solution you provided is fine.

**Comment by LitoMore** [comment]
Hi @SamVerschueren, please review this pull request https://github.com/SamVerschueren/resolve-alfred-prefs/pull/8, it's the latest one. My pull request will close #5 and #6 when it merged.

> It would be nice if we could retrieve the version from the plist file to always provide it. That way, we could enforce engines`.alfred setting` in package.json so it fails to install if you are installing an Alfred 4 workflow for Alfred 3. (@SamVerschueren)

Currently, we don't have a good way to get the version for Alfred 4 or later.

**Comment by karlsander** [comment]
hey, sorry for not keeping up with the discussion. Does #8 already superseed this PR or should I get it into shape to be merged with the new suggested api?

**Comment by LitoMore** [comment]
@karlsander My new PR #8 will fix this issue. You could suspend or close current issue. I used [GitHub issue keywords](https://help.github.com/en/articles/closing-issues-using-keywords) to close this issue. When my PR merges, your PR will also be closed at the same time.

**Comment by SamVerschueren** [comment]
No problem @karlsander, let's continue in #8

**Pull request closed without merging**